src/gui/widgets/combobox.cpp

```diff
@@ -0,0 +1,464 @@
+											bool& handled,
+											const point& coordinate)
+{
+	DBG_GUI_E << get_control_type() << "[" << id() << "]: " << event << ".";
+
+	if(dragging_) {
+		handle_mouse_selection(coordinate, false);
+	} else {
+		update_mouse_cursor();
+	}
+
+	handled = true;
+}
+
+void combobox::signal_handler_left_button_down(const event::ui_event event,
+												bool& handled)
+{
+	DBG_GUI_E << LOG_HEADER << ' ' << event << ".";
+
+	/* get_x() is the left border
+	 * this->get_size().x is the size of this widget
+	 * so get_x() + this->get_size().x is the right border
+	 * ICON_SIZE is the size of the icon. */
+
+	unsigned x = get_x() + this->get_size().x;
+	unsigned mx = get_mouse_position().x;
+
+	if ((mx <= x) && (mx >= x-ICON_SIZE)) {
+		// If a button has a retval do the default handling.
+		dialogs::drop_down_menu droplist(this, values_, selected_, false);
+
+		if(droplist.show()) {
+			const int selected = droplist.selected_item();
+
+			// Safety check. If the user clicks a selection in the dropdown and moves their mouse away too
+			// quickly, selected_ could be set to -1. This returns in that case, preventing crashes.
+			if(selected < 0) {
+				return;
+			}
+
+			set_selected(selected, true);
+		}
+	} else {
+		/*
+		 * Copied from the base class to see how we can do inheritance with the new
+		 * system...
+		 */
+		get_window()->keyboard_capture(this);
+		get_window()->mouse_capture();
+
+		handle_mouse_selection(get_mouse_position(), true);
+	}
+
+	handled = true;
+}
+
+void combobox::signal_handler_left_button_up(const event::ui_event event,
+											  bool& handled)
+{
+	DBG_GUI_E << LOG_HEADER << ' ' << event << ".";
+
+	dragging_ = false;
+	handled = true;
+}
+
+void
+combobox::signal_handler_left_button_double_click(const event::ui_event event,
+												   bool& handled)
+{
+	DBG_GUI_E << LOG_HEADER << ' ' << event << ".";
+
+	select_all();
+	handled = true;
+}
+
+// }---------- DEFINITION ---------{
+
+combobox_definition::combobox_definition(const config& cfg)
+	: styled_widget_definition(cfg)
+{
+	DBG_GUI_P << "Parsing combobox " << id;
+
+	load_resolutions<resolution>(cfg);
+}
+
+combobox_definition::resolution::resolution(const config& cfg)
+	: resolution_definition(cfg)
+	, text_x_offset(cfg["text_x_offset"])
+	, text_y_offset(cfg["text_y_offset"])
+{
+	// Note the order should be the same as the enum state_t in combobox.hpp.
+	state.emplace_back(VALIDATE_WML_CHILD(cfg, "state_enabled", _("Missing required state for editable text box")));
+	state.emplace_back(VALIDATE_WML_CHILD(cfg, "state_disabled", _("Missing required state for editable text box")));
+	state.emplace_back(VALIDATE_WML_CHILD(cfg, "state_focused", _("Missing required state for editable text box")));
+	state.emplace_back(VALIDATE_WML_CHILD(cfg, "state_hovered", _("Missing required state for editable text box")));
+}
+
+// }---------- BUILDER -----------{
+
+namespace implementation
```

Comment:
I think widget implementation documentation should go above here.

Reply:
Updated.